### DISCUSSION TITLE
Configuration to save output in notebooks

### DISCUSSION THREAD

**Discussion by jimmywan** [vega/altair#2561] [Q&A] [answered]
I hadn't used the 4.X releases until recently and while I appreciate the way [data_server](https://github.com/altair-viz/altair_data_server) allows for a much friendlier iterative analysis/development cycle, I'm confused as to the recommended workflow.
When it comes time to save/share my notebook with others, how should I save my results?
I've got some teammates with whom I'd like to share the results without a running jupyter environment and I had simply pointed them at my committed notebook in GH in the past.

I thought I could accomplish this by disabling the following:
```python
# alt.data_transformers.enable('data_server')
```
and/or adding this back in:
```python
alt.renderers.enable('notebook')
```

I'm interested in best practices for both jupyter lab and jupyter notebook.

**Comment by joelostblom** [comment, marked as answer]
You can use `alt.renderers.enable('mimetype')` [as mentioned on this page](https://altair-viz.github.io/user_guide/display_frontends.html#altair-s-renderer-framework):

> (default prior to Altair 4.0): Output a vega-lite specific mimetype that can be interpreted by appropriate frontend extensions to display charts. This also outputs a PNG representation of the plot, which is useful to view plots offline or on platforms that don’t support rendering vegaspecs, such as GitHub.

**Reply by jimmywan** [reply]
Prior to altair 4, I used a mix of notebook/lab and was able to obtain in-browser and offline rendering (as png?) that was auto-saved to my notebook.

With altair 4, when operating in Jupyter Notebook, it looks like I now have to display the chart twice, once with the html renderer and another time with notebook renderer in order to get live and offline rendering. Is that correct?

**Reply by jakevdp** [reply]
Prior to altair 4, the mimetype renderer was the default, and so in jupyterlab PNGs were saved in the notebook by default.

Unfortunately, the mimetype renderer doesn't work on jupyter notebook, and only works correctly on jupyterlab if the right vega lab-extension version is installed.

Altair 4 moved to a different default renderer that works automatically on most frontends, but unfortunately does not embed PNGs.

If you want to display charts *and* embed pngs, then:
- when using JupyterLab, you need to use the `"mimetype"` renderer (the old default) and ensure that your labextension supports the right vega-lite version
- when using Jupyter Notebook, use the `"notebook"` renderer and ensure that your nbextension supports the correct vega-lite version

The links Joel provided should give you hints on how to check and install those versions. Hope that helps

**Reply by jimmywan** [reply]
Thanks @jakevdp and @joelostblom . I think I had all of my labextension/nbextensions setup correctly, but I couldn't get the notebook renderer to do what I wanted. Easy enough to just switch to Jupyter Lab full-time and the `mimetype` renderer worked great.
I think the only reason why I had used notebook to begin with were the Jupyter Lab teething issues (many unrelated to Altair) from a few years back which appear to have all been resolved.
